### ISSUE TITLE
Support `readonly` type operator to allow correct definition of `Object.freeze`

### ISSUE THREAD

**Issue by spiffytech** [microsoft/TypeScript#10725]
**TypeScript Version:**  2.0@RC

**Code**

``` ts
// Current definition of `Object.freeze`:
// freeze<T>(o: T): T;

const o = Object.freeze({a: 1});
o.a = 2;  // Should throw type error
console.log(o);  // {a: 1}

```

**Expected behavior:**
Frozen object has `readonly` properties
**Actual behavior:** 
Type system allows me to write to `o`'s properties

**Comment by mhegazy** [comment]
There is no readonly type operator; i.e. `freeze<T>(o: T):  readonly T;`. so there is no way to do this at the time being.

We have discussed adding a readonly type operator that would recursively mark all properties as readonly.

**Comment by mhegazy** [comment]
updating the title.

**Comment by deregtd** [comment]
In our huge Typescript app, we're using a data storage mechanism that keeps complicated class-based types inside data stores.  The rest of the app consumes these objects from the stores, but there's always a risk of developers trying to modify the objects that come out of the stores, which ruins delta checking across the app.  We could use object.freeze(), but that plays havoc with JIT.

Instead, we maintain duplicate deep readonly versions of all of our model objects.  The stores distribute readonly versions, and if anyone actually needs to update them, they first clone the readonly object back into an impl object, modify it, and then shove it back in the store.  The store takes it in and stores it as a readonly version.

The general data access pattern works fantastically.  There's no javascript overhead from object.freeze(), and our engineers are unable to do terrible things with our data models (outside of <any> casting, of course...)  The big issue is just that we have a ton of these deep readonly models that we're manually maintaining, and also having to have comments to maintain which methods are readonly-accessible.  We're used to just using "const" in C++ to solve this problem, both on the accessors (from stores) and for the readonly-safe methods in a class, so this is definitely a step backward, and has introduced some engineer errors by not keeping the two versions in sync.

At least we have readonly at all, it was a huge boon to us to get that. :)  I'm just hoping for a true deep "const"-style readonly someday.

**Comment by asfernandes** [comment]
Should not this be closed after #12114 and `Readonly<T>`?

**Comment by deregtd** [comment]
It depends if there's any hope of a const-style.  We haven't bothered using ReadOnly<> because, to be frank, it's worthless to us.  We need deep readonly, so we have just created basically deep readonly types as copies of normal types.  I don't really understand how shallow readonly helps, unless all of your interfaces are just a list of primitives, without any deep classes on them.

**Comment by asfernandes** [comment]
@deregtd I agree with you on worthless of shallow Readonly. But Object.freeze is shallow too and is already changed per this issue.

**Comment by deregtd** [comment]
Fair enough.  Should I just open a new one asking for true const?

**Comment by asfernandes** [comment]
Isn't what #11535 (referenced in this issue) is already asking for?

**Comment by deregtd** [comment]
Kinda.  He's asking to mark interfaces as full readonly.  I want to have a basic set of interfaces, and then be able to mark certain usages of them as const, and have it be deep C++-style const, without having to make a complete second copy of all of them.  All the attached bugs seem to basically be solved by the partial types business.

**Comment by asfernandes** [comment]
C++-style const does not do what you say when using references and pointers (and that's how js/ts works), for example:

```cpp
class Parent
{
public:
    const Child* child;
};

const Parent* parent = ...;
parent->child->iCanMutateThis = true;
```

I think the idea should not be about have a const making a variable readonly, but a form of deeply apply something like Readonly<>. See my #13214 and the referenced issue #12424. Unfortunately, seems #12424 is more about theory than something practical.

**Comment by deregtd** [comment]
I understand the reference issue, but there's still a pretty easy way to do a better version of it in TS.

class SubBlah {
  c: string;
}

class Blah {
   a: string;
   b: SubBlah;

  const getA(): string {
    return this.a;
  }

  const getC(): string {
    return this.b.c;
  }

  setC(val: string) {
    this.b.c = val;
  }
}

function DoStuffConst(d: const Blah) {
  const a = d.getA();  //Works
  const b = d.getC();  //Works
  d.setC('a');  //Doesn't compile
}

function DoStuff(d: Blah) {
  const a = d.getA();  //Works
  const b = d.getC();  //Works
  d.setC('a');  //Works now
}

**Comment by gcanti** [comment]
@deregtd what about

```ts
type DeepReadonly<T> = { readonly [K in keyof T]: DeepReadonly<T[K]> }

type T = DeepReadonly<{
  a: {
    b: number
  }
}>

const x: T = { a: { b: 1 } }

x.a.b = 2 // error Cannot assign to 'b' because it is a constant or a read-only property.
```

**Comment by deregtd** [comment]
Interesting.  Will have to play with that concept.  That still doesn't get us there (can't have immutable methods and mutable methods, so anything with methods is hosed), but it might help some of our simpler method-less models that we currently have copies of (RO and RW).  Thanks!

**Comment by asfernandes** [comment]
@gcanti this does not work correctly for embedded arrays.

**Comment by gcanti** [comment]
@asfernandes neither for tuples. Without mapped conditional types is pretty hard to write conditional logic. This is the best result I got so far (with some bad hacks)

```ts
declare global {
  interface Array<T> {
    _type: T
    _kind: 'Array'
  }
  interface Object {
    _kind: 'Object'
  }
}

export type Readonlyable = Array<any> | Object

export type DeepReadonlyObject<T> = {
  readonly [K in keyof T]: DeepReadonlyObject<T[K]>
}

export type OneLevelReadonly<T extends Readonlyable> = {
  Array: ReadonlyArray<Readonly<T['_type']>>,
  Object: { readonly [K in keyof T]: DeepReadonlyObject<T[K]> }
}[T['_kind']]

export type DeepReadonly<T extends { [key: string]: Readonlyable }> = { readonly [K in keyof T]: OneLevelReadonly<T[K]> }

type T = DeepReadonly<{
  a: {
    b: {
      c: number
      // bar: Foo // this leads to Element implicitly has an 'any' type
      // baz: Array<number> // this leads to Element implicitly has an 'any' type
    }
  },
  d: Array<number>,
  e: Array<{ foo: number }>,
  f: Array<{ foo: number, bar: { bax: number } }>, // this doesn't work, bax is not readonly
  g: Array<{ foo: number, bar: Array<number> }> // this doesn't work, bar is not readonly
  // the previuos 2 cases must be handled with another type definition like in h
  h: Array<Foo>,
  i: Foo
}>

type Foo = DeepReadonly<{ foo: number, bar: Array<number> }>

declare var x: T

x.a.b.c = 2 // ok, error
x.d[0] = 2 // ok, error
x.e[0].foo = 2 // ok, error
x.f[0].foo = 2 // ok, error
x.f[0].bar.bax = 2 // NO error
x.g[0].foo = 2 // ok, error
x.g[0].bar[0] = 2 // NO error
x.i.bar = 2 // ok, error
```

**Comment by MichaelTontchev** [comment]
Would like to add my voice to the request for a deep, recursive readonly. Would be super useful and make immutability very good in TypeScript.

**Comment by lukescott** [comment]
I can currently use this:
```
export type Freeze<T> = {
	readonly [P in keyof T]: T[P]
}
```
Which allows me to set each property as readonly:
```ts
type Foo = Freeze<{
  one: string
  two: string
}>
var foo: Foo = {one:"one", two:"two"}
foo.one = "ONE" // error
```
The issue I have with this is `type Foo` isn't named, and I lose other benefits of using an interface over a type def.

I would like to be able to do something like this:
```ts
readonly interface Foo {
  one: string
  two: string
}
```

It could be a deep readonly, but I would be ok with it being shallow. Ultimately I'm trying to avoid doing this:

```ts
interface Foo {
  readonly one: string
  readonly two: string
}
```

Because if someone adds another property, it may not be clear that `Foo` is supposed to be treated as an immutable type.

**Comment by skishore** [comment]
Very nice, @gcanti! I think I was able to extend your example to be closer to working!

```
declare global {
  interface Array<T> {
    _type: T
    _kind: 'Array'
  }
  interface Object {
    _kind: 'Object'
  }
}

export type Readonlyable = Array<any> | Object

export type OneLevelReadonly<T extends Readonlyable> = {
  Array: ReadonlyArray<DeepReadonly<T['_type']>>,
  Object: DeepReadonly<T>, // This is the key difference between my version and yours.
}[T['_kind']]

export type DeepReadonly<T extends { [key: string]: Readonlyable }> = { readonly [K in keyof T]: OneLevelReadonly<T[K]> }

type T = DeepReadonly<{
  a: {
    b: {
      c: number
      bar: Foo // this leads to Element implicitly has an 'any' type
      baz: Array<number> // this leads to Element implicitly has an 'any' type
    }
    foo: Foo,
  },
  d: Array<number>,
  e: Array<{ foo: number }>,
  f: Array<{ foo: number, bar: { bax: number } }>, // this doesn't work, bax is not readonly
  g: Array<{ foo: number, bar: Array<number> }> // this doesn't work, bar is not readonly
  // the previuos 2 cases must be handled with another type definition like in h
  h: Array<Foo>,
  i: Foo
}>

type Foo = { foo: number, bar: Array<number> };

declare var x: T
// All have type number!
const a = x.a.b.c;
const b = x.a.b.bar.bar[0];
const c = x.a.b.baz[0];
const d = x.a.foo.foo;
const e = x.a.foo.bar[0];

x.a.b.c = 2 // ok, error
x.d[0] = 2 // ok, error
x.e[0].foo = 2 // ok, error
x.f[0].foo = 2 // ok, error
x.f[0].bar.bax = 2 // ok, error
x.g[0].foo = 2 // ok, error
x.g[0].bar[0] = 2 // ok, error
x.i.bar = 2 // ok, error
```

Note that things will still break if I make Foo a DeepReadonly object. However, with this method, we can at least use DeepReadonly as a replacement for C++ style "const" on functions, since that use case only ever depends on one level of DeepReadonly being available. That's the main way I wanted to use it, so I'm very pleased. Nice trick with the "_type" / "_kind" way to go from Array<T> -> T!

**Comment by skishore** [comment]
One more go: I think I've ironed out a lot of problems. It now works for primitives, for ES6 Map and Set, and for some pretty complex nested structures. Best of all, all you need to do is export the "Const" type and use it, and type T can be assigned to type Const<T> without issue.

```
// An attempt at implementing const-correctness in Typescript.

declare global {
  interface Array<T> {_kind: 'array', _t1: T, _t2: never}
  interface Map<K,V> {_kind: 'map', _t1: K, _t2: V}
  interface Object {_kind: 'object', _t1: never, _t2: never}
  interface Set<T> {_kind: 'set', _t1: T}
};

export type Const<T extends Array<any> | Map<any,any> | Object | Set<any>> = {
  array: ConstArray<T & {length: number}>,
  map: ReadonlyMap<Readonly<T['_t1']>,Readonly<T['_t2']>>;
  object: ConstObject<T>,
  set: ReadonlySet<Readonly<T['_t1']>>,
}[T['_kind']];

type ConstArray<T extends ArrayLike<any>> = {
  0: ReadonlyArray<Const<T[0]>>,
  1: {
    0: Readonly<[Const<T[0]>]>,
    1: {
      0: Readonly<[Const<T[0]>,Const<T[1]>]>,
      1: {
        0: Readonly<[Const<T[0]>,Const<T[1]>,Const<T[2]>]>,
        1: Readonly<[Const<T[0]>,Const<T[1]>,Const<T[2]>,Const<T[3]>]>,
      }[TupleHasIndex<T,'3'>]
    }[TupleHasIndex<T,'2'>]
  }[TupleHasIndex<T,'1'>]
}[TupleHasIndex<T,'0'>];

type ConstObject<T extends {[key: string]: any}> =
  {readonly [K in keyof T]: Const<T[K]>};

type TupleHasIndex<A, I extends string> =
  ({[K in keyof A]: '1'} & {[key: string]: '0'})[I];

// Example usages. Note that the only interface you need is Const.
// You can check that things are working correctly by using the Typescript server's
// GetType call on these values. You'll see that they reduce to simple types with
// "readonly" and Readonly[Array|Map|Set] where appropriate.
declare const a: Const<number>;
declare const b: Const<string>;
declare const c: Const<boolean>;
declare const d: Const<Map<number,{a: boolean, b: string}>>;
declare const e: Const<Set<string>>;
declare const x: Const<number[]>;
declare const y: Const<[number, string, boolean, {a: number, b: string}]>;
declare const z: Const<{a: number[], b: {c: string, d: string[]}[]}>;

type MyType = [number[], {a: string, b: boolean}];
const value: MyType = [[1, 2, 3], {a: 'test', b: true}];
const f: Const<MyType> = value;
``` 

I'll wrap this up in a library and make it available.

**Comment by tycho01** [comment]
@skishore: interesting, that looks like something that'd be nice for `typelevel-ts` / `typical` as well :D, though I hope we could get to an iterative solution not limited to x layers.
Coincidentally, there's been a few attempts at #12424 as well recently.

I think one of the tough parts right now is still how to deal with objects with string/number index types. That is to say, `{ a: 123 }` may work, but it's currently tough to handle both that and `{ a: 123; [k: string]: number }` in the same algorithm... that one requires obtaining its index type (done by accessing `o[string]`), but that just errors when done on an object without it.

I'd imagined that #6606 would finally allow making a check to confirm whether there is such an index, but in my current PR for that (#17961) found that this isn't actually the case. At this point I have no idea how we might go about tackling that part... which would definitely be more unfortunate if it would recursively cause information loss throughout the object structure.

**Comment by skishore** [comment]
There's no limitation for the approach above on layers - the only numeric limit is tuple length, which is a pretty standard thing in Typescript definitions that I've seen.

I also haven't seen any problems with mixed index types as you suggest - I just tried one out. Admittedly, there were a few bugs in that initial code listing. I'm ironing those out now as I apply const-correctness to one of my projects.

There is a very serious issue I've found, though. Typescript seems to drop readonly object fields at will when calling a function. This problem is unrelated to the Const type, as you can see here:

```typescript
type Record = {a: number, b: string};
const test = (record: Record) => void {};
const record: Readonly<Record> = {a: 1, b: 'test'}; /* fields are readonly */
test(record); /* No compile error! */
```

So...what's the point of making things const-correct if you can accidentally cast it away by calling a function?

**Comment by tycho01** [comment]
Yeah, your approach wouldn't error on input with an index. To my understanding though, the output wouldn't retain them, but rather discard them, while the ideal behavior would be to both retain the indices ~~while making them read-only as well~~. Edit: apparently `{ [readonly k: string]: number }` yields `An index signature parameter cannot have an accessibility modifier`. Ouch, that seems... not ideal.

I know @gcanti used definitions of this style (explicitly do each arity), yeah -- I may need to look into how that could be resolved again. I suppose this workaround probably had to do with some occasional constraint issues of the iteration approach, I forgot.

The behavior you mentioned does seem weird...

**Comment by skishore** [comment]
You want `{readonly [k: string]: number}`, I think.

The readonly assignment check is a known issue:
https://github.com/Microsoft/TypeScript/issues/13347 The current behavior
(allowing readonly interfaces to be passed as mutable parameters) was by
design to avoid having to propagate readonly everywhere in an existing
codebase. But having const-correctness hold throughout the codebase is the
whole point of const for me. I would consider working on the
"enforceReadonlyAssignment" flag if there was consensus that that would be
a good feature, but without it, I don't see the value in this type
annotation.

**Comment by tycho01** [comment]
```
declare let a: { readonly [k: string]: number }; // does this work?
a.b = 2; // nope, passes
```
... yeah, `readonly` seems to have its share of problems.

That said though, my PR at #17785 may be relevant as a workaround by inferring accurate typings where reasonable (parameters / `const`), meaning your `[1, 2, 3]` gets type `[1, 2, 3]` instead of `number[]`. The beauty of that is it pretty much disables assigning anything new as well (unless it's the same value).

**Comment by ChuckJonas** [comment]
also struggling with the deep immutability issues.. Would love to see a solution

**Comment by renevaessen** [comment]
What am I missing with this:

type DeepReadonly<T> = {
    readonly [P in keyof T]:  DeepReadonly<T[P]>;
 };

http://tinyurl.com/ydbga82j

**Comment by tycho01** [comment]
@renevaessen:
- arrays
- tuples
- string indices
- symbols

**Comment by niieani** [comment]
Here's a deep readonly that also makes Arrays deeply readonly:[Playground](https://www.typescriptlang.org/play/#src=type%20Example%20%3D%20%7B%0D%0A%20%20a%3A%20string%2C%0D%0A%20%20b%3A%20Array%3CExample%3E%2C%0D%0A%20%20c%3A%20number%2C%0D%0A%20%20d%3A%20Example%2C%0D%0A%7D%0D%0A%0D%0Adeclare%20var%20test%3A%20RecursiveReadonly%3CExample%3E%3B%0D%0A%2F%2F%20errors%20properly%3A%0D%0Atest.b%5B0%5D.b%5B0%5D.d.b%5B0%5D.a%20%3D%20''%0D%0Atest.b%5B0%5D.b%5B0%5D.d.b%5B0%5D.b%20%3D%20%5B%5D%0D%0Atest.b%5B0%5D.b%5B0%5D.d.b%5B0%5D.c%20%3D%203%0D%0Atest.b%5B0%5D.b%5B0%5D.d.b%5B0%5D.d%20%3D%20test%0D%0A%0D%0Adeclare%20var%20test2%3A%20RecursiveReadonly%3Cstring%3E%3B%0D%0Atest2%20%3D%20test2%20as%20string%0D%0A%0D%0Adeclare%20var%20test3%3A%20RecursiveReadonly%3Cnumber%3E%3B%0D%0Atest3%20%3D%20test3%20as%20number%0D%0A%0D%0Adeclare%20var%20test4%3A%20RecursiveReadonly%3CSet%3Cstring%3E%3E%3B%0D%0Atest4%20%3D%20test4%20as%20Set%3Cstring%3E%0D%0A%0D%0A%0D%0A%2F%2F%20definition%3A%0D%0Atype%20RecursiveReadonly%3CT%3E%20%3D%20If%3C%0D%0A%20%20IsArrayType%3CT%3E%2C%0D%0A%20%20MapArrayType%3CT%20%26%20List%3C%7B%7D%3E%3E%2C%0D%0A%20%20MapScalarOrObjectType%3CT%3E%0D%0A%3E%3B%0D%0A%0D%0Atype%20MapScalarOrObjectType%3CT%3E%20%3D%20If%3C%0D%0A%20%20IsScalarOrInstance%3CT%3E%2C%0D%0A%20%20T%2C%0D%0A%20%20MapObjectWithPossibleArrays%3CT%3E%0D%0A%3E%3B%0D%0A%0D%0Atype%20MapObjectWithPossibleArrays%3CT%3E%20%3D%20%7B%0D%0A%20%20readonly%20%5BP%20in%20keyof%20T%5D%3A%20If%3C%0D%0A%20%20%20%20IsArrayType%3CT%5BP%5D%3E%2C%0D%0A%20%20%20%20MapArrayType%3CT%5BP%5D%20%26%20List%3C%7B%7D%3E%3E%2C%0D%0A%20%20%20%20MapScalarOrObjectType%3CT%5BP%5D%3E%0D%0A%20%20%3E%0D%0A%7D%3B%0D%0A%0D%0Atype%20MapArrayType%3CArr%20extends%20List%3CT%3E%2C%20T%20%3D%20Arr%5B-1%5D%3E%20%3D%20ReadonlyArray%3C%0D%0A%20%20MapObjectWithPossibleArrays%3CT%3E%0D%0A%3E%3B%0D%0A%0D%0A%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%0D%0A%2F%2F%20generic%20utils%20(some%20from%20https%3A%2F%2Fgithub.com%2Ftycho01%2Ftypical%2F)%3A%0D%0Atype%20False%20%3D%20%220%22%3B%0D%0Atype%20True%20%3D%20%221%22%3B%0D%0Atype%20Bool%20%3D%20True%20%7C%20False%3B%0D%0Atype%20If%3CCond%20extends%20Bool%2C%20Then%2C%20Else%3E%20%3D%20%7B%201%3A%20Then%3B%200%3A%20Else%20%7D%5BCond%5D%3B%0D%0A%0D%0Atype%20Obj%3CT%3E%20%3D%20%7B%20%5Bk%3A%20string%5D%3A%20T%20%7D%3B%0D%0A%0D%0Atype%20And%3CA%20extends%20Bool%2C%20B%20extends%20Bool%3E%20%3D%20(%7B%201%3A%20%7B%201%3A%20%221%22%20%7D%20%26%20Obj%3C%220%22%3E%20%7D%20%26%20Obj%3C%0D%0A%20%20Obj%3C%220%22%3E%0D%0A%3E)%5BA%5D%5BB%5D%3B%0D%0A%0D%0Atype%20UnionHasKey%3CUnion%20extends%20string%2C%20K%20extends%20string%3E%20%3D%20(%7B%0D%0A%20%20%5BS%20in%20Union%5D%3A%20%221%22%0D%0A%7D%20%26%0D%0A%20%20Obj%3C%220%22%3E)%5BK%5D%3B%0D%0A%0D%0Atype%20Indeterminate%3CT%20extends%20string%3E%20%3D%20And%3C%0D%0A%20%20UnionHasKey%3CT%2C%20%220%22%3E%2C%0D%0A%20%20UnionHasKey%3CT%2C%20%221%22%3E%0D%0A%3E%3B%0D%0A%0D%0Atype%20Not%3CT%20extends%20Bool%3E%20%3D%20%7B%20%221%22%3A%20%220%22%3B%20%220%22%3A%20%221%22%20%7D%5BT%5D%3B%0D%0A%0D%0Atype%20Determinate%3CT%20extends%20Bool%3E%20%3D%20Not%3CIndeterminate%3CT%3E%3E%3B%0D%0A%0D%0Atype%20DefinitelyYes%3CT%20extends%20Bool%3E%20%3D%20And%3CT%2C%20Determinate%3CT%3E%3E%3B%0D%0A%0D%0Atype%20UnionContained%3CT%20extends%20string%2C%20U%20extends%20string%3E%20%3D%20DefinitelyYes%3C%0D%0A%20%20(%7B%20%5BP%20in%20U%5D%3A%20%221%22%20%7D%20%26%20Obj%3C%220%22%3E)%5BT%20%7C%20U%5D%0D%0A%3E%3B%0D%0A%0D%0Atype%20UnionEmpty%3CT%20extends%20string%3E%20%3D%20And%3C%0D%0A%20%20UnionContained%3CT%2C%20%22foo%22%3E%2C%0D%0A%20%20UnionContained%3CT%2C%20%22bar%22%3E%0D%0A%3E%3B%0D%0A%0D%0Atype%20UnionToObject%3CKeys%20extends%20string%3E%20%3D%20%7B%20%5BK%20in%20Keys%5D%3A%20K%20%7D%3B%0D%0A%0D%0Atype%20Keyed%3CT%3E%20%3D%20%7B%20%5BK%20in%20keyof%20T%5D%3A%20K%20%7D%3B%0D%0A%0D%0Atype%20KeyedSafe%3CT%3E%20%3D%20Keyed%3CT%3E%20%26%20Obj%3Cnever%3E%3B%0D%0A%0D%0Atype%20IntersectionUnions%3CBig%20extends%20string%2C%20Small%20extends%20string%3E%20%3D%20KeyedSafe%3C%0D%0A%20%20UnionToObject%3CSmall%3E%0D%0A%3E%5BBig%5D%3B%0D%0Atype%20UnionsOverlap%3CBig%20extends%20string%2C%20Small%20extends%20string%3E%20%3D%20Not%3C%0D%0A%20%20UnionEmpty%3CIntersectionUnions%3CBig%2C%20Small%3E%3E%0D%0A%3E%3B%0D%0A%0D%0Atype%20DiffUnion%3CT%20extends%20string%2C%20U%20extends%20string%3E%20%3D%20(%7B%20%5BP%20in%20T%5D%3A%20P%20%7D%20%26%0D%0A%20%20%7B%20%5BP%20in%20U%5D%3A%20never%20%7D%20%26%20%7B%20%5Bk%3A%20string%5D%3A%20never%20%7D)%5BT%5D%3B%0D%0A%0D%0Atype%20ObjectHasKey%3CO%20extends%20%7B%7D%2C%20K%20extends%20string%3E%20%3D%20UnionHasKey%3Ckeyof%20O%2C%20K%3E%3B%0D%0A%0D%0Atype%20ArrayPrototypeProperties%20%3D%20DiffUnion%3C%0D%0A%20%20keyof%20Array%3Cany%3E%2C%0D%0A%20%20%22toString%22%20%7C%20%22toLocaleString%22%0D%0A%3E%3B%0D%0Atype%20IsArrayType%3CT%3E%20%3D%20DefinitelyYes%3CObjectHasKey%3CT%2C%20ArrayPrototypeProperties%3E%3E%3B%0D%0A%0D%0A%2F**%0D%0A%20*%20false%20for%3A%20undefined%2C%20interfaces%0D%0A%20*%20true%20for%3A%20Array%2C%20boolean%2C%20number%2C%20string%2C%20Object%0D%0A%20*%2F%0D%0Atype%20IsScalarOrInstance%3CT%3E%20%3D%20ObjectHasKeySafe%3CT%2C%20%22valueOf%22%3E%3B%0D%0Atype%20ObjectHasKeySafe%3CO%2C%20K%20extends%20string%3E%20%3D%20UnionsOverlap%3Ckeyof%20O%2C%20K%3E%3B%0D%0A%0D%0Ainterface%20List%3CT%3E%20%7B%0D%0A%20%20readonly%20%5Bn%3A%20number%5D%3A%20T%3B%0D%0A%7D).

Let me know if you find any issues with it. Thanks to @tycho01 for some of the utility types.

**Comment by tycho01** [comment]
@niieani: nice one. seems `MapObjectWithPossibleArrays` is duplicating a bit from `RecursiveReadonly`, have you tried defining it in terms of that?
also /cc @pelotom

**Comment by niieani** [comment]
Indeed it seems to work just with:
```
type MapObjectWithPossibleArrays<T> = {
  readonly [P in keyof T]: RecursiveReadonly<T[P]>
};
```
I think during development I've had some issues with TS complaining about it, but after a number of refactorings it seems it works now as above.

**Comment by AlexGalays** [comment]
@niieani 

That's some serious type manipulation haha

Does it cover all known test cases so far?

**Comment by tycho01** [comment]
@AlexGalays: the following we can't preserve yet, limitations not of his type but of our set of operators we can use today:
- tuples
- string indices
- symbols

Note that at #12424 they're similarly trying to lift a key based operator to become 'deep', in their case partial/? rather than readonly.

**Comment by AlexGalays** [comment]
@tycho01 Strange, string indices (or Records) appear to work just fine: They're read-only and the value type you read follows the index declaration.

**Comment by tycho01** [comment]
@AlexGalays: the tough part is ideally we'd wanna preserve explicitly defined indices throughout operations, but I think we don't have a way to check for them.
What we can do is try to access them, but if we do while they weren't there, the type just errors, which we cannot currently 'catch'.

**Comment by pelotom** [comment]
I just want to submit that ideally a solution here could also solve another problem which is one of my biggest gripes with TypeScript: [unnecessary type widening](https://github.com/Microsoft/TypeScript/issues/20195). Since `readonly` fields should not be widened, it seems like literals passed to `Object.freeze` should not have their types widened:

```ts
const o = Object.freeze({ x: 3, y: 'hello' });
// o: { readonly x: number; readonly y: string }
// but would ideally be
// o: { readonly x: 3; readonly y: 'hello' }
```

In the linked issue I proposed a `readonly` term operator for this purpose, but it would actually be preferable if it were possible to express a type for `Object.freeze` that could accomplish this, since then I could give the same type to a function which did nothing but return the original object, eliminating the runtime overhead if all I want is the static check. Thus being able to type `Object.freeze` such that it inferred the type ` { readonly x: 3; readonly y: 'hello' }` in the above example would be a strictly more powerful feature.

**Comment by tycho01** [comment]
I think the `DeepReadOnly` implementation in #21316 should mean this issue is now resolved.
All that seems left is to have it added to `lib.d.ts` and have the `Object.freeze` definition adjusted accordingly, though technically they seem outside of what was asked here.

**Comment by ikokostya** [comment]
One note: `DeepReadOnly` implementation in https://github.com/Microsoft/TypeScript/pull/21316 removes function all function properties, but `Object.freeze` doesn't.

**Comment by jeremejevs** [comment]
Not sure why function props are singled out in #21316. Perhaps to showcase the usage of `NonFunctionPropertyNames<T>`? Regardless, unless I'm missing something, this ought do the trick:

```ts
type DeepReadonly<T> =
  T extends any[] ? DeepReadonlyArray<T[number]> :
  T extends object ? DeepReadonlyObject<T> :
  T;

interface DeepReadonlyArray<T> extends ReadonlyArray<DeepReadonly<T>> {}

type DeepReadonlyObject<T> = T &
  { readonly [P in keyof T]: DeepReadonly<T[P]> };
```

[Playground.](http://www.typescriptlang.org/play/#src=type%20DeepReadonly%3CT%3E%20%3D%0D%0A%20%20%20%20T%20extends%20any%5B%5D%20%3F%20DeepReadonlyArray%3CT%5Bnumber%5D%3E%20%3A%0D%0A%20%20%20%20T%20extends%20object%20%3F%20DeepReadonlyObject%3CT%3E%20%3A%0D%0A%20%20%20%20T%3B%0D%0A%0D%0Ainterface%20DeepReadonlyArray%3CT%3E%20extends%20ReadonlyArray%3CDeepReadonly%3CT%3E%3E%20%7B%20%7D%0D%0A%0D%0Atype%20DeepReadonlyObject%3CT%3E%20%3D%20T%20%26%0D%0A%20%20%7B%20readonly%20%5BP%20in%20keyof%20T%5D%3A%20DeepReadonly%3CT%5BP%5D%3E%20%7D%3B%0D%0A%0D%0Atype%20Hybrid%20%3D%20%7B%0D%0A%20%20()%3A%20string%3B%0D%0A%20%20prop%3F%3A%20string%3B%0D%0A%7D%3B%0D%0A%0D%0Aconst%20x%3A%20Hybrid%20%3D%20()%20%3D%3E%20'foo'%3B%0D%0Ax.prop%20%3D%20'x'%3B%0D%0A%0D%0Aconst%20y%3A%20Hybrid%20%3D%20()%20%3D%3E%20'bar'%3B%0D%0Ay.prop%20%3D%20'y'%3B%0D%0A%0D%0Ainterface%20A%20%7B%0D%0A%20%20num%3A%20number%3B%0D%0A%20%20func%3A%20()%20%3D%3E%20string%2C%0D%0A%20%20hybrid%3A%20Hybrid%3B%0D%0A%7D%0D%0A%0D%0Atype%20B%20%3D%20DeepReadonly%3CA%3E%3B%0D%0A%0D%0Aconst%20a%3A%20A%20%3D%20%7B%20num%3A%20123%2C%20func%3A%20()%20%3D%3E%20'foo'%2C%20hybrid%3A%20x%20%7D%3B%0D%0Aa.num%20%3D%20456%3B%0D%0Aa.func%20%3D%20()%20%3D%3E%20'bar'%3B%0D%0Aa.func%20%3D%20x%3B%0D%0Aa.func.prop%20%3D%20'a'%3B%20%2F%2F%20Error%0D%0Aa.hybrid%20%3D%20y%3B%0D%0Aa.hybrid.prop%20%3D%20'a'%3B%0D%0A%0D%0Aconst%20b%3A%20B%20%3D%20a%3B%0D%0Ab.num%20%3D%20789%3B%20%2F%2F%20Error%0D%0Ab.func%20%3D%20()%20%3D%3E%20'baz'%3B%20%2F%2F%20Error%0D%0Ab.hybrid%20%3D%20x%3B%20%2F%2F%20Error%0D%0Ab.hybrid.prop%20%3D%20'b'%3B%20%2F%2F%20Error%0D%0A%0D%0Aconst%20c%3A%20A%20%3D%20b%3B%0D%0Ac.num%20%3D%20789%3B%0D%0A)

Note the union in `DeepReadonlyObject<T>`, it takes care of both regular functions and hybrid interfaces-functions, like this one:

```ts
interface A {
  (): string;
  prop: number;
}
```

Just in case, I don't know TS that well, so there may be side effects.

**Comment by karlhorky** [comment]
This can be closed now that the `freeze` method uses the `Readonly` and `ReadonlyArray` generic utility types, I take it?

https://github.com/microsoft/TypeScript/blob/b36c8a06902ca47da1568167991bd0bc28ced80e/src/lib/es5.d.ts#L195-L211

**Comment by deregtd** [comment]
Maybe for the original post, yes, but then I will have to break my follow up post into a separate one for deep read only since that is only one level deep. :)

**Comment by karlhorky** [comment]
Yep, should probably close the original issue and open a followup asking for a built-in [Utility Type](https://www.typescriptlang.org/docs/handbook/utility-types.html) for `DeepReadonly`.

Since the implementation above and in https://github.com/microsoft/TypeScript/pull/21316#issue-164138025 is written by ahejlsberg, maybe it's not too far away from a the quality for inclusion?

cc @spiffytech

**Comment by tycho01** [comment]
@karlhorky 
> maybe it's not too far away from the quality for inclusion?

well type recursion is not deemed a supported use of the language so far :/

**Comment by lazytype** [comment]
I propose the issue be closed now. Read-only array types have been around for a year, which means the following works as expected. @DanielRosenwasser @RyanCavanaugh 
```ts
type DeepReadonly<T> = { readonly [K in keyof T]: DeepReadonly<T[K]> };

type Foo = { a: { b: { c: [ { x: 'y' } ] } } };
const foo: Foo = { a: { b: { c: [ { x: 'y' } ] } } };

type ReadonlyFoo = DeepReadonly<Foo>;

const readonlyFoo: ReadonlyFoo = foo;

// All type errors
readonlyFoo.a = readonlyFoo.a;
readonlyFoo.a.b = readonlyFoo.a.b;
readonlyFoo.a.b.c = readonlyFoo.a.b.c;
readonlyFoo.a.b.c[0] = readonlyFoo.a.b.c[0];
readonlyFoo.a.b.c[0].x = readonlyFoo.a.b.c[0].x;
```

**Comment by MartinJohns** [comment]
I'd keep it open, but adjust the issue instead. Given the mention of `Object.freeze` it's clear that he (and I as well) seek to represent immutability in the type-system, which currently is not supported.

**Comment by RyanCavanaugh** [comment]
I agree this is effectively solved for the use cases in OP and related posts. New issues welcomed for further scenarios.

**Comment by ExE-Boss** [comment]
See also:
- <https://github.com/microsoft/TypeScript/issues/30680>
- <https://github.com/microsoft/TypeScript/issues/42862>
